pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -75,20 +78,47 @@ func GetDefaultCircuitBreakerThresholds(direction model.TrafficDirection) *v2Clu
 func (configgen *ConfigGeneratorImpl) BuildClusters(env *model.Environment, proxy *model.Proxy, push *model.PushContext) ([]*apiv2.Cluster, error) {
 	clusters := make([]*apiv2.Cluster, 0)
 
-	switch proxy.Type {
-	case model.SidecarProxy:
-		instances, err := env.GetProxyServiceInstances(proxy)
-		if err != nil {
-			log.Errorf("failed to get service proxy service instances: %v", err)
-			return nil, err
+	instances, err := env.GetProxyServiceInstances(proxy)
+	if err != nil {
+		log.Errorf("failed to get service proxy service instances: %v", err)
+		return nil, err
+	}
+
+	// compute the proxy's locality. See if we have a CDS cache for that locality.
+	// If not, compute one.
+	locality := proxy.Locality
+	if locality == nil {
+		// Get the locality from the proxy's service instances.
+		// We expect all instances to have the same locality. So its enough to look at the first instance
+		if len(instances) > 0 {
+			locality = util.ConvertLocality(instances[0].Endpoint.Locality)
 		}
+	}
 
+	switch proxy.Type {
+	case model.SidecarProxy:
 		sidecarScope := proxy.SidecarScope
 		recomputeOutboundClusters := true
 		if configgen.CanUsePrecomputedCDS(proxy) {
-			if sidecarScope != nil && sidecarScope.XDSOutboundClusters != nil {
-				clusters = append(clusters, sidecarScope.XDSOutboundClusters...)
+			if sidecarScope != nil && sidecarScope.CDSOutboundClusters != nil {
+				// NOTE: We currently only cache & update the CDS output for NoProxyLocality
+				clusters = append(clusters, sidecarScope.CDSOutboundClusters[util.NoProxyLocality]...)
```

Comment:
Why use the NoProxyLocality cache even with locality?

Reply:
Because eventually I want to be able to use the logic from this commit: https://github.com/istio/istio/pull/11381/commits/b9c739ebef793ac650849c1516658121727813ad#diff-f507fb55b7edf4a03d3c7e9aebf64a1eR109